cdap-app-fabric/src/main/java/co/cask/cdap/master/environment/k8s/AbstractServiceMain.java

```diff
@@ -71,16 +69,20 @@
 import javax.annotation.Nullable;
 
 /**
- * The abstract base class for writing various CDAP master service main classes.
+ * The abstract base class for writing various service main classes.
+ *
+ * @param <T> type of options supported by the service.
  */
-public abstract class AbstractServiceMain extends DaemonMain {
+public abstract class AbstractServiceMain<T extends EnvironmentOptions> extends DaemonMain {
 
   private static final Logger LOG = LoggerFactory.getLogger(AbstractServiceMain.class);
 
   private final List<Service> services = new ArrayList<>();
   private final List<AutoCloseable> closeableResources = new ArrayList<>();
-  private Injector injector;
-  private MasterEnvironment masterEnv;
+  protected MasterEnvironment masterEnv;
```

Comment:
It's better not to expose non final field directly to child class. Either use getter method, or have these provided through method parameter when calling into abstract method (more preferred).